### PR TITLE
[MOB-21131] Display Proposition api to support multiple propositions.

### PR DESCRIPTION
**Pull request by akhiljain1907**
## Description
This PR adds an api "trackDisplayedOffers" which takes list of offers as an argument and sends display interaction event to Experience Edge.

~Through this additional api, we now support display event to be called on multiple propositions which can be send in batches and not per offer (read: proposition) and so we've moved generateInteractionXDM and trackWithData methods from Offer class to public class.~

This api supports calling display interaction event on multiple offers. Since a proposition could contain more than one offers and our generateInteractionXDM method takes list of propositions, we first extract list of unique propositions from passed offers and create new proposition with relevant offer ids to be tracked. We have also moved generateInteractionXDM and trackWithData methods to OptimizeTrackingUtils from Offer+Tracking extension to reuse them within the module.
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: Sources/AEPOptimize/Optimize+Tracking.swift [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+import AEPCore
+import AEPServices
+import Foundation
+
+@objc
+public extension Optimize {
+    /// This API dispatches an event for the Edge extension to send an Experience Event to the Edge network with the display interaction data for list of offers passed.
+    ///
+    /// - Parameter offers: An array of offer.
+    static func trackDisplayedOffers(for offers: [Offer]) {
```
Comment (spoorthipujariadobe) [inline Sources/AEPOptimize/Optimize+Tracking.swift:21]: Would it be possible to keep the API names similar across platforms? Android implementation seems to be an extension function on List<Offer> called `displayed`
https://github.com/adobe/aepsdk-optimize-android/pull/92/files#diff-8288c041a5231846e1dbc0c7f03e9c4b0c48990cfdf2fc21cd807a02ed969ee3R24

File: Sources/AEPOptimize/Optimize+Tracking.swift [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+                "items": relevantOffers.map { offer in
+                    [
+                        "id": offer.id,
+                        "schema": offer.schema,
+                        "data": [
+                            "id": offer.id,
+                            "type": offer.type.rawValue,
+                            "content": offer.content,
+                            "language": offer.language,
+                            "characteristics": offer.characteristics
```
Comment (spoorthipujariadobe) [inline Sources/AEPOptimize/Optimize+Tracking.swift:48]: Is it possible to add the entire offer object here instead of serializing it? We have missed adding some offer attributes like etag and score here, which can be avoided if we passing in the offer object as is and let its `encode` method convert it to?

Comment (spoorthipujariadobe) [reply Sources/AEPOptimize/Optimize+Tracking.swift:48]: Looking at [this](https://github.com/adobe/aepsdk-optimize-ios/pull/112/files) again, we only need the `offer.id`. So if you want to just add that instead of all the extra fields, that could work too

File: Sources/AEPOptimize/Optimize+Tracking.swift [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+            trackWithData(xdmData)
+        }
+    }
+
+    /// Creates a dictionary containing XDM formatted data for `Experience Event - Proposition Interactions` field group from the given list of propositions and for the provided event type.
+    ///
+    /// - Parameter propositions: An array of optimize propositions.
+    /// - Parameter eventType: The Experience Event event type for the proposition interaction.
+    /// - Returns A dictionary containing XDM data for the propositon interactions.
+    static func generateInteractionXdm(for propositions: [OptimizeProposition], for eventType: String) -> [String: Any]? {
```
Comment (spoorthipujariadobe) [inline Sources/AEPOptimize/Optimize+Tracking.swift:73]: Why does this need to be exposed as public API? Can we make it private?

File: Sources/AEPOptimize/Optimize+Tracking.swift [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+        ]
+        return xdmData
+    }
+
+    /// Dispatches the track propositions request event with type `EventType.optimize` and source `EventSource.requestContent` and given proposition interactions data.
+    ///
+    /// No event is dispatched if the input xdm data is `nil`.
+    ///
+    /// - Parameter xdmData: A dictionary containing XDM data for the propositon interactions.
+    static func trackWithData(_ xdmData: [String: Any]?) {
```
Comment (spoorthipujariadobe) [inline Sources/AEPOptimize/Optimize+Tracking.swift:113]: Same question here. Why does this need to be public?

File: Sources/AEPOptimize/Optimize+Tracking.swift (hunk 3 above)
Comment (akhiljain1907) [reply Sources/AEPOptimize/Optimize+Tracking.swift]: I thought we need to provide these methods as public api. Since it's not required, I have created an internal enum which has these methods. They are not marked as private as they being used at multiple places but can be accessed only within the module.

File: Sources/AEPOptimize/Optimize+Tracking.swift (hunk 4 above)
Comment (akhiljain1907) [reply Sources/AEPOptimize/Optimize+Tracking.swift]: Moved this to an internal enum.

File: Sources/AEPOptimize/Optimize+Tracking.swift (hunk 2 above)
Comment (akhiljain1907) [reply Sources/AEPOptimize/Optimize+Tracking.swift]: Yes, i have removed extra data and we can keep only offer id.

File: Sources/AEPOptimize/Optimize+Tracking.swift (hunk 1 above)
Comment (akhiljain1907) [reply Sources/AEPOptimize/Optimize+Tracking.swift]: Sure, I have renamed the api as per suggestion.

File: Tests/AEPOptimizeTests/UnitTests/OptimizePublicAPITests.swift [hunk 5]
```diff
@@ -549,6 +549,194 @@ class OptimizePublicAPITests: XCTestCase {
...
+        }
+        
+        // test
+        Optimize.displayed(for: [offer1, offer2, offer3])
+        
+        // verify
+        wait(for: [expectation], timeout: 2)
+    }
+    
+    func testGenerateInteractionXdm_multiplePropositions() throws {
```
Comment (spoorthipujariadobe) [inline Tests/AEPOptimizeTests/UnitTests/OptimizePublicAPITests.swift:687]: nit: can we move this and the following tests to a separate `OptimizeTrackingUtilsTests` file?

Comment (spoorthipujariadobe) [reply Tests/AEPOptimizeTests/UnitTests/OptimizePublicAPITests.swift:687]: Would be good to add a few more tests here to cover cases like empty propositions and all offers belonging to the same proposition. You can check the Android PR for more cases
https://github.com/adobe/aepsdk-optimize-android/pull/92/files#diff-710393a6e27a3c8cd7cb485acc24b9df4631442f2c7efc5a5aa7b7ed24686b38

**Review by spoorthipujariadobe — COMMENTED** [review]
Logic looks good. Would be nice to have someone more familiar with iOS review it as well, maybe @ishwetansh ?

File: Sources/AEPOptimize/OptimizeProposition.swift [hunk 6]
```diff
@@ -60,7 +60,7 @@ public class OptimizeProposition: NSObject, Codable {
         try container.encode(id, forKey: .id)
         try container.encode(scope, forKey: .scope)
         try container.encode(AnyCodable.from(dictionary: scopeDetails), forKey: .scopeDetails)
-        try container.encode(offers, forKey: .items)
+        try container.encode(items, forKey: .items)
```
Comment (ishwetansh) [inline Sources/AEPOptimize/OptimizeProposition.swift:63]: Please confirm if we need items or offers to be encoded here

File: Sources/AEPOptimize/OptimizeTrackingUtils.swift [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+ OF ANY KIND, either express or implied. See the License for the specific language
+ governing permissions and limitations under the License.
+ */
+
+import AEPCore
+import AEPServices
+import Foundation
+
+/// Internal helper enum for tracking functionality
+enum OptimizeTrackingUtils {
```
Comment (ishwetansh) [inline Sources/AEPOptimize/OptimizeTrackingUtils.swift:18]: I don't see the use case of using an enum here, it is generally used to define and handle finite cases.
can we please use a struct here.

File: Sources/AEPOptimize/OptimizeTrackingUtils.swift [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+                OptimizeConstants.JsonKeys.DECISIONING_PROPOSITIONS_SCOPE: proposition.scope,
+                OptimizeConstants.JsonKeys.DECISIONING_PROPOSITIONS_SCOPEDETAILS: proposition.scopeDetails,
+                OptimizeConstants.JsonKeys.DECISIONING_PROPOSITIONS_ITEMS: proposition.offers.map { offer in
+                    [
+                        OptimizeConstants.JsonKeys.DECISIONING_PROPOSITIONS_ITEMS_ID: offer.id
+                    ]
+                }
+            ]
+            propositionDetailsData.append(propositionMap)
+        }
```
Comment (ishwetansh) [inline Sources/AEPOptimize/OptimizeTrackingUtils.swift:39]: We can use a transformable `map` function here to build the dictionary instead of traditional loop.

`let propositionDetailsData: [[String: Any]] = propositions.map { proposition in
    return [
        OptimizeConstants.JsonKeys.DECISIONING_PROPOSITIONS_ID: proposition.id,
        OptimizeConstants.JsonKeys.DECISIONING_PROPOSITIONS_SCOPE: proposition.scope,
        OptimizeConstants.JsonKeys.DECISIONING_PROPOSITIONS_SCOPEDETAILS: proposition.scopeDetails,
        OptimizeConstants.JsonKeys.DECISIONING_PROPOSITIONS_ITEMS: proposition.offers.map { offer in
            return [
                OptimizeConstants.JsonKeys.DECISIONING_PROPOSITIONS_ITEMS_ID: offer.id
            ]
        }
    ]
}
`

File: Tests/AEPOptimizeTests/UnitTests/Propositions+OptimizeTests.swift [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+ Unless required by applicable law or agreed to in writing, software distributed under
+ the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ OF ANY KIND, either express or implied. See the License for the specific language
+ governing permissions and limitations under the License.
+ */
+
+
+import Foundation
+
+class Propositions_OptimizeTests {
```
Comment (ishwetansh) [inline Tests/AEPOptimizeTests/UnitTests/Propositions+OptimizeTests.swift:17]: I guess, we don't usually use an _ in class names,
can we use `PropositionsOptimizeTests`

File: Sources/AEPOptimize/OptimizeTrackingUtils.swift (hunk 7 above)
Comment (akhiljain1907) [reply Sources/AEPOptimize/OptimizeTrackingUtils.swift:18]: since we are using only static methods here, it';s recommended to use enum. 
FYI when i use struct, i get this warning - Convenience Type Violation: Types used for hosting only static members should be implemented as a caseless enum to avoid instantiation (convenience_type)

File: Sources/AEPOptimize/OptimizeTrackingUtils.swift (hunk 8 above)
Comment (akhiljain1907) [reply Sources/AEPOptimize/OptimizeTrackingUtils.swift]: Done

File: Tests/AEPOptimizeTests/UnitTests/Propositions+OptimizeTests.swift (hunk 9 above)
Comment (akhiljain1907) [reply Tests/AEPOptimizeTests/UnitTests/Propositions+OptimizeTests.swift]: I previously kept this file in Utils folder of Unit test where this naming convention was followed. But now i have moved these test data in a new file PropositionsTestData which is kept in TestHelpers folder.

File: Tests/AEPOptimizeTests/UnitTests/OptimizePublicAPITests.swift [hunk 10]
```diff
@@ -682,6 +870,7 @@ class OptimizePublicAPITests: XCTestCase {
             XCTAssertEqual(propositions.count, 1)
             XCTAssertEqual(propositions.first?.id, proposition.id)
             XCTAssertEqual(propositions.first?.scope, proposition.scope)
+
```
[the remark this answers is not in the document]
Comment (akhiljain1907) [reply Tests/AEPOptimizeTests/UnitTests/OptimizePublicAPITests.swift:685]: Done

File: Tests/AEPOptimizeTests/UnitTests/OptimizePublicAPITests.swift (hunk 5 above)
Comment (akhiljain1907) [reply Tests/AEPOptimizeTests/UnitTests/OptimizePublicAPITests.swift]: I have moved testGenerateInteractionXdm_multiplePropositions() in OptimizePropositionTests as that's were relevant generate xdm tests were present. 
Also i have added a new test file OptimizeTrackingTests which contains test for new displayed api.

**Review by ishwetansh — APPROVED** [review]
